### PR TITLE
FEATURE NMH-130: Incremental nlp website build [skip-test]

### PR DESCRIPTION
**Pull request by KshitizGIT**
Jira Ticket [NMH-130](https://johnsnowlabs.atlassian.net/browse/NMH-130)

@pabla , 
- Support for full build and incremental build
- In full build, sites and elastic search indexes are rebuilt from start
- Some required environment variables on Jenkins
       - ELASTICSEARCH_URL
       - ELASTICSEARCH_ACCESS_TOKEN
       - GITHUB_TOKEN
       - SPARK_NLP_REPO

### CODE REVIEW

File: docs/_plugins/search_index.rb [hunk 1]
```diff
@@ -432,6 +423,7 @@ def is_latest?(group, model)
 end
 
 Jekyll::Hooks.register :site, :post_render do |site|
+  is_incremental = site.config['incremental']
   force_reindex = editions_changed?(editions)
```
Comment (pabla) [inline docs/_plugins/search_index.rb:427]: Looks like force_reindex is not being used.
If we see a new edition, `Healthcare NLP 4.2`, for example, we should re-index all Healthcare NLP models and mark them as compatible with `Healthcare NLP 4.2`. The same situation is with models.json: we should update the `compatible_models` field.

**Review by pabla — COMMENTED** [review]
Is it possible to move Dockerfile, Jenkinsfile, and docker-compose.yml to the docs directory? They are related to the website only.

File: docs/_plugins/search_index.rb (hunk 1 above)
Comment (KshitizGIT) [reply docs/_plugins/search_index.rb:427]: So, if if the edition changes, its best to do a full build. Am I right, @pabla ?

Comment (pabla) [reply docs/_plugins/search_index.rb:427]: Yes. Or update affected models.

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,21 @@
+FROM ruby:2.7
```
[the remark this answers is not in the document]
Comment (KshitizGIT) [reply Dockerfile]: Done

**Comment by KshitizGIT** [comment]
> Is it possible to move Dockerfile, Jenkinsfile, and docker-compose.yml to the docs directory? They are related to the website only.

Done

File: docs/_plugins/search_index.rb (hunk 1 above)
Comment (KshitizGIT) [reply docs/_plugins/search_index.rb]: Will trigger a full incremental build from the Jenkins when newer editions are found. This is done by removing the backup files `docs/.jekyll-metadata`, `docs/models.json`, `docs/references.json`, `docs/benchmarking.json`

File: docs/_plugins/search_index.rb [hunk 3]
```diff
@@ -79,7 +79,10 @@ def editions_changed?(local_editions)
   if res.is_a?(Net::HTTPSuccess)
     data = JSON.parse(res.body)
     remote_editions = data['meta']['aggregations']['editions']
-    return !local_editions.to_set.subset?(remote_editions.to_set)
+    outdated_editions = ['Spark NLP 2.1', 'Healthcare NLP 2.0']
```
Comment (pabla) [inline docs/_plugins/search_index.rb:82]: I see the same `outdated_editions` in another method. Maybe it makes sense to extract it to a constant?

File: docs/_plugins/search_index.rb [hunk 4]
```diff
@@ -481,34 +480,77 @@ def is_latest?(group, model)
...
+          bulk_indexer.index(id, model)
         end
       end
     end
   end
   bulk_indexer.execute
 
-  if client
-    # Also delete models whose  name, language, edition, spark version were modified
+  if client and not is_incremental
```
Comment (pabla) [inline docs/_plugins/search_index.rb:493]: We have to delete models from the Elasticsearch index in the incremental builds as well.

Comment (KshitizGIT) [reply docs/_plugins/search_index.rb:493]: Removed models will appear via this hooks 
Jekyll::Hooks.register :clean, :on_obsolete do |files|
I have implemented deleted models here for incremental build

File: docs/_plugins/search_index.rb (hunk 3 above)
Comment (KshitizGIT) [reply docs/_plugins/search_index.rb]: Used a constant variables for both

File: docs/_plugins/search_index.rb [hunk 5]
```diff
@@ -6,6 +6,8 @@
 require 'elasticsearch'
 require 'nokogiri'
 
+$outdated_editions = ['Spark NLP 2.1', 'Healthcare NLP 2.0']
```
Comment (pabla) [inline docs/_plugins/search_index.rb:9]: Why not `OUTDATED_EDITIONS`?

Comment (KshitizGIT) [reply docs/_plugins/search_index.rb]: Yeah I updated it.. I doesn't work with `outdated_editions` . But works with the Ruby constant `OUTDATED_EDITIONS`

**Review by pabla — APPROVED** [review]

**Pull request merged**